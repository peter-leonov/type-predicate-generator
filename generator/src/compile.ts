import ts from "typescript";
import { TypeGuardGenerator } from "./generator";
import { assert, unimplemented } from "./helpers";
import { TypeModel, typeToModel } from "./model";
import { TypeScriptError } from "./errors";

/**
 * This is required for both `X[]` and `Array<X>` to work properly.
 * Otherwise `checker.isArrayType()` is not gonna work.
 */
const libFileContent = `/// <reference no-default-lib="true"/>
interface Boolean {}
interface Function {}
interface CallableFunction {}
interface NewableFunction {}
interface IArguments {}
interface Number { toExponential: any; }
interface Object {}
interface RegExp {}
interface String { charAt: any; }
interface Array<T> { length: number; [n: number]: T; }
interface ReadonlyArray<T> {}
declare const console: { log(msg: any): void; };`;

export function newVFSProgram(source: string): ts.Program {
  const fileName = "/source.ts";

  const files = new Map<string, string>();
  files.set(fileName, source);

  const libFileName = "/lib/lib.d.ts";
  files.set(libFileName, libFileContent);

  const compilerHost: ts.CompilerHost = {
    fileExists: (fileName) => files.has(fileName),
    getSourceFile: (fileName, options) => {
      const content = files.get(fileName);
      assert(content != undefined, "source files cannot be missing");
      return ts.createSourceFile(fileName, content, options);
    },
    getDefaultLibFileName: () => "/lib/lib.d.ts",
    writeFile: unimplemented,
    getCurrentDirectory: () => "/",
    getDirectories: () => [],
    getCanonicalFileName: (f) => f.toLowerCase(),
    getNewLine: () => "\n",
    useCaseSensitiveFileNames: () => false,
    readFile: (path) => files.get(path),
  };

  return ts.createProgram(
    [fileName],
    {
      target: ts.ScriptTarget.ESNext,
      module: ts.ModuleKind.ESNext,
      strict: true,
      noEmit: true,
      isolatedModules: true,
      // don't autoload all the types from the node_modules/@types
      types: [],
    },
    compilerHost
  );
}

export function ensureNoErrors(program: ts.Program) {
  const allDiagnostics = ts.getPreEmitDiagnostics(program);
  if (allDiagnostics.length != 0) {
    allDiagnostics.forEach((diagnostic) => {
      var message = ts.flattenDiagnosticMessageText(
        diagnostic.messageText,
        "\n"
      );
      if (!diagnostic.file) {
        console.error(message);
        return;
      }
      var { line, character } =
        diagnostic.file.getLineAndCharacterOfPosition(
          diagnostic.start!
        );
      console.error(
        `${diagnostic.file.fileName} (${line + 1},${
          character + 1
        }): ${message}`
      );
    });

    throw new TypeScriptError(
      "ts.getPreEmitDiagnostics() found some issues listed in the console / stderr"
    );
  }
}

function hasExportModifier(
  node:
    | ts.TypeAliasDeclaration
    | ts.InterfaceDeclaration
    | ts.EnumDeclaration
): boolean {
  return (
    node.modifiers?.some(
      (node) => node.kind === ts.SyntaxKind.ExportKeyword
    ) ?? false
  );
}

export function sourceFileToDeclarationSymbols(
  checker: ts.TypeChecker,
  sourceFile: ts.SourceFile
): ts.Symbol[] {
  const symbols: ts.Symbol[] = [];

  // Walk the tree to search for types
  ts.forEachChild(sourceFile, (node) => {
    if (
      ts.isTypeAliasDeclaration(node) ||
      ts.isInterfaceDeclaration(node) ||
      ts.isEnumDeclaration(node)
    ) {
      if (!hasExportModifier(node)) return;
      let symbol = checker.getSymbolAtLocation(node.name);
      assert(symbol, "a node declaration must have a symbol");
      symbols.push(symbol);
    }
  });

  return symbols;
}

export function symbolsToModels(
  checker: ts.TypeChecker,
  symbols: ts.Symbol[]
): TypeModel[] {
  return symbols.map((symbol) => typeToModel(checker, symbol));
}

export function sourceFileToModels(
  checker: ts.TypeChecker,
  sourceFile: ts.SourceFile
): TypeModel[] {
  return symbolsToModels(
    checker,
    sourceFileToDeclarationSymbols(checker, sourceFile)
  );
}

export function generateFullFileBodyForAllTypes(
  checker: ts.TypeChecker,
  sourceFile: ts.SourceFile,
  importFrom: string
): ts.Statement[] {
  const generator = new TypeGuardGenerator();

  for (const model of sourceFileToModels(checker, sourceFile)) {
    generator.addRootTypeGuardFor(model);
  }

  return generator.getFullFileBody(importFrom);
}

export function sourceFileToString(
  sourceFile: ts.SourceFile
): string {
  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
  });

  return printer.printFile(sourceFile);
}
