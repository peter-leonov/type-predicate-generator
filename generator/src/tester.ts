import ts, { factory } from "typescript";
import {
  combineInvalid,
  combineValid,
  invalidValue,
  modelToCombinator,
} from "./combinator";
import { type TypeModel } from "./model";
import { assert } from "./helpers";

export function modelsToTests(
  predicatesFileName: string,
  models: TypeModel[]
): ts.Statement[] {
  const predicateNames: string[] = [];
  const tests = [];
  for (const model of models) {
    const typeName = model.options.aliasName;
    assert(
      typeName,
      "root type model has to have the type aliasName defined"
    );
    const predicateName = `is${typeName}`;
    predicateNames.push(predicateName);
    tests.push(...modelToTests(typeName, predicateName, model));
  }

  return [
    getImports(predicatesFileName, predicateNames),
    defineInvalidValue(),
    ...tests,
  ];
}

export function modelToTests(
  typeName: string,
  predicateName: string,
  model: TypeModel
): ts.Statement[] {
  const combinator = modelToCombinator(model);

  const valids = combineValid(combinator);
  const validsName = `valid_${typeName}`;

  const invalids = combineInvalid(combinator);
  const invalidsName = `invalid_${typeName}`;

  return [
    valuesVar(validsName, valids),
    valuesVar(invalidsName, invalids),
    describeItFor(typeName, validsName, invalidsName, predicateName),
  ];
}

function getImports(
  sourceFileName: string,
  imports: string[]
): ts.Statement {
  return factory.createImportDeclaration(
    undefined,
    factory.createImportClause(
      false,
      undefined,

      factory.createNamedImports(
        imports.map((className) => {
          return factory.createImportSpecifier(
            false,
            undefined,
            factory.createIdentifier(className)
          );
        })
      )
    ),
    factory.createStringLiteral(sourceFileName),
    undefined
  );
}

function valuesVar(name: string, values: unknown[]): ts.Statement {
  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier(name),
          undefined,
          undefined,
          factory.createArrayLiteralExpression(
            valuesToExpression(values),
            true
          )
        ),
      ],
      ts.NodeFlags.Const
    )
  );
}

const invalidValueVarName = "invalidValue";

function defineInvalidValue(): ts.Statement {
  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier(invalidValueVarName),
          undefined,
          factory.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
          factory.createCallExpression(
            factory.createIdentifier("Symbol"),
            undefined,
            [factory.createStringLiteral("invalidValue")]
          )
        ),
      ],
      ts.NodeFlags.Const
    )
  );
}

export const invalidValueToken =
  "80D79A5A-5D38-4302-BA04-F5AC748C1464";

function valuesToExpression(values: unknown[]): ts.Expression[] {
  const res: ts.Expression[] = [];

  const seen = new Set();
  for (const value of values) {
    const json = JSON.stringify(value, (_, v) => {
      return v === invalidValue ? invalidValueToken : v;
    });
    if (seen.has(json)) {
      continue;
    }
    seen.add(json);
    const file = ts.parseJsonText("valueToJSONExpression", json);
    const statement = file.statements[0];
    assert(statement, "must have a single statement");
    res.push(statement.expression);
  }

  return res;
}

/**
 * Dirty hack to avoid wirting a JS value to Node parser just to cover
 * the invalid value symbol.
 */
export function hydrateInvalidValueToken(code: string): string {
  return code.replaceAll(
    `"${invalidValueToken}"`,
    invalidValueVarName
  );
}

function describeItFor(
  describeName: string,
  validsName: string,
  invalidsName: string,
  predicateName: string
): ts.Statement {
  return factory.createExpressionStatement(
    factory.createCallExpression(
      factory.createIdentifier("describe"),
      undefined,
      [
        factory.createStringLiteral(describeName),
        factory.createArrowFunction(
          undefined,
          undefined,
          [],
          undefined,
          factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
          factory.createBlock(
            [
              factory.createExpressionStatement(
                factory.createCallExpression(
                  factory.createCallExpression(
                    factory.createPropertyAccessExpression(
                      factory.createIdentifier("it"),
                      factory.createIdentifier("for")
                    ),
                    undefined,
                    [factory.createIdentifier(validsName)]
                  ),
                  undefined,
                  [
                    factory.createStringLiteral("valid"),
                    factory.createArrowFunction(
                      undefined,
                      undefined,
                      [
                        factory.createParameterDeclaration(
                          undefined,
                          undefined,
                          factory.createIdentifier("value"),
                          undefined,
                          undefined,
                          undefined
                        ),
                      ],
                      undefined,
                      factory.createToken(
                        ts.SyntaxKind.EqualsGreaterThanToken
                      ),
                      factory.createBlock(
                        [
                          factory.createExpressionStatement(
                            factory.createCallExpression(
                              factory.createPropertyAccessExpression(
                                factory.createCallExpression(
                                  factory.createIdentifier("expect"),
                                  undefined,
                                  [
                                    factory.createCallExpression(
                                      factory.createIdentifier(
                                        predicateName
                                      ),
                                      undefined,
                                      [
                                        factory.createIdentifier(
                                          "value"
                                        ),
                                      ]
                                    ),
                                  ]
                                ),
                                factory.createIdentifier("toBe")
                              ),
                              undefined,
                              [factory.createTrue()]
                            )
                          ),
                        ],
                        true
                      )
                    ),
                  ]
                )
              ),
              factory.createExpressionStatement(
                factory.createCallExpression(
                  factory.createCallExpression(
                    factory.createPropertyAccessExpression(
                      factory.createIdentifier("it"),
                      factory.createIdentifier("for")
                    ),
                    undefined,
                    [factory.createIdentifier(invalidsName)]
                  ),
                  undefined,
                  [
                    factory.createStringLiteral("invalid"),
                    factory.createArrowFunction(
                      undefined,
                      undefined,
                      [
                        factory.createParameterDeclaration(
                          undefined,
                          undefined,
                          factory.createIdentifier("value"),
                          undefined,
                          undefined,
                          undefined
                        ),
                      ],
                      undefined,
                      factory.createToken(
                        ts.SyntaxKind.EqualsGreaterThanToken
                      ),
                      factory.createBlock(
                        [
                          factory.createExpressionStatement(
                            factory.createCallExpression(
                              factory.createPropertyAccessExpression(
                                factory.createCallExpression(
                                  factory.createIdentifier("expect"),
                                  undefined,
                                  [
                                    factory.createCallExpression(
                                      factory.createIdentifier(
                                        predicateName
                                      ),
                                      undefined,
                                      [
                                        factory.createIdentifier(
                                          "value"
                                        ),
                                      ]
                                    ),
                                  ]
                                ),
                                factory.createIdentifier("toBe")
                              ),
                              undefined,
                              [factory.createFalse()]
                            )
                          ),
                        ],
                        true
                      )
                    ),
                  ]
                )
              ),
            ],
            true
          )
        ),
      ]
    )
  );
}
