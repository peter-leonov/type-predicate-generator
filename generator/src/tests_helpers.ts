import ts from "typescript";
import { factory } from "typescript";
import assert from "assert";
import {
  ensureNoErrors,
  newVFSProgram,
  sourceFileToDeclarationSymbols,
} from "./compile";

export function compile(
  source: string
): [ts.TypeChecker, ts.Symbol[]] {
  const program = newVFSProgram(source);
  ensureNoErrors(program);

  const sourceFile = program.getSourceFiles()[1];
  assert(
    sourceFile,
    "the program must have 2 source files, one for the lib and one for the actual test code"
  );

  const checker = program.getTypeChecker();

  // This finds the LAST type definition in the source.
  const symbols = sourceFileToDeclarationSymbols(checker, sourceFile);
  assert(
    symbols.length,
    "at least one symbol must be present in the test case"
  );
  return [checker, symbols];
}

export function printNodes(nodes: ts.Statement[]): string {
  const resultFile = factory.updateSourceFile(
    ts.createSourceFile(
      "guards.ts",
      "",
      ts.ScriptTarget.Latest,
      /*setParentNodes*/ false,
      ts.ScriptKind.TS
    ),
    nodes
  );

  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
  });
  return printer.printFile(resultFile);
}
