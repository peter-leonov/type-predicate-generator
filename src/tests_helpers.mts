import ts from "typescript";
import { unimplemented } from "./helpers.mts";
import { ok } from "assert";

export function compile(
  source: string
): [ts.TypeChecker, ts.Type, ts.Symbol] {
  const fileName = "/source.ts";

  const sourceFiles = new Map<string, ts.SourceFile>();
  sourceFiles.set(
    fileName,
    ts.createSourceFile(fileName, source, ts.ScriptTarget.ESNext)
  );

  const compilerHost: ts.CompilerHost = {
    fileExists: (fileName) => sourceFiles.has(fileName),
    getSourceFile: (fileName) => sourceFiles.get(fileName),
    getDefaultLibFileName: () => "lib.d.ts",
    writeFile: unimplemented,
    getCurrentDirectory: () => "/",
    getDirectories: () => [],
    getCanonicalFileName: (f) => f.toLowerCase(),
    getNewLine: () => "\n",
    useCaseSensitiveFileNames: () => false,
    readFile: unimplemented,
  };

  const program = ts.createProgram(
    [fileName],
    {
      target: ts.ScriptTarget.ESNext,
      module: ts.ModuleKind.ESNext,
    },
    compilerHost
  );

  const sourceFile = program.getSourceFiles()[0];
  ok(sourceFile);

  const checker = program.getTypeChecker();

  let symbol: ts.Symbol | undefined;
  ts.forEachChild(sourceFile, (node) => {
    if (!ts.isTypeAliasDeclaration(node)) {
      return;
    }

    symbol = checker.getSymbolAtLocation(node.name);
    ok(symbol);
  });

  ok(symbol);
  const type = checker.getDeclaredTypeOfSymbol(symbol);
  return [checker, type, symbol];
}
