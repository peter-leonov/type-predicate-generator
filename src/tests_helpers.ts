import ts from "typescript";
import { factory } from "typescript";
import { unimplemented } from "./helpers";
import { ok } from "assert";

/**
 * This is required for both `X[]` and `Array<X>` to work properly.
 * Otherwise `checker.isArrayType()` is not gonna work.
 */
const libFileContent = `/// <reference no-default-lib="true"/>
interface Boolean {}
interface Function {}
interface CallableFunction {}
interface NewableFunction {}
interface IArguments {}
interface Number { toExponential: any; }
interface Object {}
interface RegExp {}
interface String { charAt: any; }
interface Array<T> { length: number; [n: number]: T; }
interface ReadonlyArray<T> {}
declare const console: { log(msg: any): void; };`;

export function compile(
  source: string
): [ts.TypeChecker, ts.Type, ts.Symbol] {
  const fileName = "/source.ts";

  const files = new Map<string, string>();
  files.set(fileName, source);

  const libFileName = "/lib/lib.d.ts";
  files.set(libFileName, libFileContent);

  const compilerHost: ts.CompilerHost = {
    fileExists: (fileName) => files.has(fileName),
    getSourceFile: (fileName, options) => {
      const content = files.get(fileName);
      ok(content);
      return ts.createSourceFile(fileName, content, options);
    },
    getDefaultLibFileName: () => "/lib/lib.d.ts",
    writeFile: unimplemented,
    getCurrentDirectory: () => "/",
    getDirectories: () => [],
    getCanonicalFileName: (f) => f.toLowerCase(),
    getNewLine: () => "\n",
    useCaseSensitiveFileNames: () => false,
    readFile: (path) => files.get(path),
  };

  const program = ts.createProgram(
    [fileName],
    {
      target: ts.ScriptTarget.ESNext,
      module: ts.ModuleKind.ESNext,
      strict: true,
      noEmit: true,
      isolatedModules: true,
      // don't autoload all the types from the node_modules/@types
      types: [],
    },
    compilerHost
  );

  const allDiagnostics = ts.getPreEmitDiagnostics(program);
  if (allDiagnostics.length != 0) {
    allDiagnostics.forEach((diagnostic) => {
      var message = ts.flattenDiagnosticMessageText(
        diagnostic.messageText,
        "\n"
      );
      if (!diagnostic.file) {
        console.error(message);
        return;
      }
      var { line, character } =
        diagnostic.file.getLineAndCharacterOfPosition(
          diagnostic.start!
        );
      console.error(
        `${diagnostic.file.fileName} (${line + 1},${
          character + 1
        }): ${message}`
      );
    });

    throw new Error(
      "ts.getPreEmitDiagnostics() found some issues listed above"
    );
  }

  const sourceFile = program.getSourceFiles()[1];
  ok(sourceFile);

  const checker = program.getTypeChecker();

  // This finds the LAST type definition in the source.
  let symbol: ts.Symbol | undefined;
  ts.forEachChild(sourceFile, (node) => {
    if (!ts.isTypeAliasDeclaration(node)) {
      return;
    }

    symbol = checker.getSymbolAtLocation(node.name);
    ok(symbol);
    // console.dir(checker.getTypeFromTypeNode(node.type));
  });

  ok(symbol);
  const type = checker.getDeclaredTypeOfSymbol(symbol);
  return [checker, type, symbol];
}

export function printNodes(nodes: ts.Statement[]): string {
  const resultFile = factory.updateSourceFile(
    ts.createSourceFile(
      "guards.ts",
      "",
      ts.ScriptTarget.Latest,
      /*setParentNodes*/ false,
      ts.ScriptKind.TS
    ),
    nodes
  );

  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
  });
  return printer.printFile(resultFile);
}
