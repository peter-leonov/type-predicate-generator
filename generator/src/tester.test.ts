import { expect, test } from "vitest";
import { hydrateInvalidValueToken, modelsToTests } from "./tester";
import { compile, printNodes } from "./tests_helpers";
import { symbolsToModels } from "./compile";

function process(code: string): string {
  const [checker, symbols] = compile(code);
  const models = symbolsToModels(checker, symbols);

  return hydrateInvalidValueToken(
    printNodes(modelsToTests("guards.ts", models))
  );
}

test("primitive", () => {
  expect(process("export type X = number")).toMatchSnapshot();
});

test("array", () => {
  expect(process("export type X = string[]")).toMatchSnapshot();
});

test("object", () => {
  expect(
    process("export type X = { a: string, b: number, c: boolean }")
  ).toMatchSnapshot();
});

test("union", () => {
  expect(
    process("export type X = string | number")
  ).toMatchSnapshot();
});

test("multiple types", () => {
  expect(
    process(`
      export type X = string | number
      export type Y = boolean | null
      `)
  ).toMatchSnapshot();
});
