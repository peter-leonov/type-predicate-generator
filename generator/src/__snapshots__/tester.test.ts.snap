// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`array 1`] = `
"import { isX } from "guards.ts";
const invalidValue: any = Symbol("invalidValue");
const valid_X = [
    [],
    ["string"]
];
const invalid_X = [
    invalidValue,
    [invalidValue]
];
describe("X", () => {
    it.for(valid_X)("valid", value => {
        expect(isX(value)).toBe(true);
    });
    it.for(invalid_X)("invalid", value => {
        expect(isX(value)).toBe(false);
    });
});
"
`;

exports[`multiple types 1`] = `
"import { isX, isY } from "guards.ts";
const invalidValue: any = Symbol("invalidValue");
const valid_X = [
    "string",
    0,
    42
];
const invalid_X = [
    invalidValue
];
describe("X", () => {
    it.for(valid_X)("valid", value => {
        expect(isX(value)).toBe(true);
    });
    it.for(invalid_X)("invalid", value => {
        expect(isX(value)).toBe(false);
    });
});
const valid_Y = [
    null,
    false,
    true
];
const invalid_Y = [
    invalidValue
];
describe("Y", () => {
    it.for(valid_Y)("valid", value => {
        expect(isY(value)).toBe(true);
    });
    it.for(invalid_Y)("invalid", value => {
        expect(isY(value)).toBe(false);
    });
});
"
`;

exports[`object 1`] = `
"import { isX } from "guards.ts";
const invalidValue: any = Symbol("invalidValue");
const valid_X = [
    { "a": "string", "b": 0, "c": true },
    { "a": "string", "b": 0, "c": false },
    { "a": "string", "b": 42, "c": true },
    { "a": "string", "b": 42, "c": false }
];
const invalid_X = [
    invalidValue,
    { "b": 0, "c": true },
    { "b": 0, "c": false },
    { "b": 42, "c": true },
    { "b": 42, "c": false },
    { "a": invalidValue, "b": 0, "c": true },
    { "a": invalidValue, "b": 0, "c": false },
    { "a": invalidValue, "b": 42, "c": true },
    { "a": invalidValue, "b": 42, "c": false },
    { "a": "string", "c": true },
    { "a": "string", "c": false },
    { "a": "string", "b": invalidValue, "c": true },
    { "a": "string", "b": invalidValue, "c": false },
    { "a": "string", "b": 0 },
    { "a": "string", "b": 0, "c": invalidValue },
    { "a": "string", "b": 42 },
    { "a": "string", "b": 42, "c": invalidValue }
];
describe("X", () => {
    it.for(valid_X)("valid", value => {
        expect(isX(value)).toBe(true);
    });
    it.for(invalid_X)("invalid", value => {
        expect(isX(value)).toBe(false);
    });
});
"
`;

exports[`primitive 1`] = `
"import { isX } from "guards.ts";
const invalidValue: any = Symbol("invalidValue");
const valid_X = [
    0,
    42
];
const invalid_X = [
    invalidValue
];
describe("X", () => {
    it.for(valid_X)("valid", value => {
        expect(isX(value)).toBe(true);
    });
    it.for(invalid_X)("invalid", value => {
        expect(isX(value)).toBe(false);
    });
});
"
`;

exports[`union 1`] = `
"import { isX } from "guards.ts";
const invalidValue: any = Symbol("invalidValue");
const valid_X = [
    "string",
    0,
    42
];
const invalid_X = [
    invalidValue
];
describe("X", () => {
    it.for(valid_X)("valid", value => {
        expect(isX(value)).toBe(true);
    });
    it.for(invalid_X)("invalid", value => {
        expect(isX(value)).toBe(false);
    });
});
"
`;
